### ISSUE TITLE
{shutdown,connack_timeout} when benchmark

### ISSUE THREAD

**Issue by dcy** [erlio/vernemq#120]
Hi!
I already set the config:
listener.max_connections = infinity
listener.nr_of_acceptors = 1000
erlang.max_ports = 10000000
erlang.process_limit = 10000000
And use [emqtt_benchmark](https://github.com/emqtt/emqtt_benchmark)

Two console:
./emqtt_bench_sub -c 20000 -i 10 -t bench/%i -q 2
./emqtt_bench_pub -c 100 -I 10 -t bench/%i -s 25

A few minutes, it will be:
[error] [Client <0.8154.1>] CONNACK Timeout!
client 6372 EXIT: {shutdown,connack_timeout}
Topics: ["bench/%i"]
conneted: 19054
[error] [Client <0.8157.1>] CONNACK Timeout!
client 6371 EXIT: {shutdown,connack_timeout}
Topics: ["bench/%i"]
conneted: 19055
[error] [Client <0.8160.1>] CONNACK Timeout!
client 6370 EXIT: {shutdown,connack_timeout}
Topics: ["bench/%i"]
conneted: 19056

Another time, when all the clients conneted, suddenly many clients got "{shutdown, tcp_closed}".
It works with [emqttd](https://github.com/emqtt/emqttd)

**Comment by dcy** [comment]
I test again, and this time, all the clients connectd, and work well a while, the sub clients:
client 17564 EXIT: {shutdown,tcp_closed}
client 17563 EXIT: {shutdown,tcp_closed}
client 17562 EXIT: {shutdown,tcp_closed}
client 17561 EXIT: {shutdown,tcp_closed}
client 17560 EXIT: {shutdown,tcp_closed}
client 17558 EXIT: {shutdown,tcp_closed}
client 17559 EXIT: {shutdown,tcp_closed}

**Comment by ioolkos** [comment]
Thanks for your input!

Can you help us a little understanding what the scenario does?

It seems you setup a 100 publishers doing 10k msg/s total. (but to what topics?)
and 20k subscribers, but to what topics?

**Comment by dcy** [comment]
I have not look the code of emqtt_benchmark, I just take the tool to test emqttd and vernemq.

By the way, if vernemq has something like [emqttd_dashboard](https://github.com/emqtt/emqttd_dashboard) would be better.
[Demo](http://emqtt.io/docs/getstarted.html#web-dashboard)

**Comment by dcy** [comment]
For example, create 50K concurrent clients at the arrival rate of 100/sec:
./emqtt_bench_sub -c 50000 -i 10 -t bench/%i -q 2

For example, create 100 clients and each client publish messages at the rate of 100 msg/sec.
./emqtt_bench_pub -c 100 -I 10 -t bench/%i -s 256
Topic is "bench/%i"

**Comment by ioolkos** [comment]
OK, just guessing here: 
First you seem to have 20k subscribers to 20k exclusive topics. But you send messages to only 100 of those topics, so the other connections will timeout, if they don't answer to keepalive.

Will have to look into the CONNACK problem otoh.

As I said, this is just guessing atm, I'll try to verify with your benchmark.

**Comment by dcy** [comment]
Is that  the vernemq server can not handle too much client connections in the same time?

**Comment by ioolkos** [comment]
No, at least 100 connection setup per second shouldn't be any problem.

By the way: an MQTT client can decide itself what it's CONNACK timeout is. I don't know what it is for your benchmark.

I found out though, that the default keepalive set is 600 seconds (that is, if you don't set any own keepalive timeout). This means your subscriber will disconnect after 5 minutes, if they never issue any MQTT control packets.

**Comment by dcy** [comment]
I use [emqtt_benchmark](https://github.com/emqtt/emqtt_benchmark) to test [emqttd](https://github.com/emqtt/emqttd), it will be fine.
But i use [emqtt_benchmark](https://github.com/emqtt/emqtt_benchmark) to test [VerneMQ](https://github.com/erlio/vernemq), it will be connack_timeout in the same situation.

**Comment by dergraf** [comment]
I guess the jobs rate limiter kicks in.

**Comment by dcy** [comment]
Thank you for your reply.
Can i make VerneMQ higher performance easily? like change some configuration.
I just decided to use VerneMQ, not emqttd, but in the benchmark, found this situation.

**Comment by dergraf** [comment]
If this is the reasonit would be a config setting away... I have to verify it and get back to you.

**Comment by dergraf** [comment]
Hi there.

OK, I did the test you mentioned in your first post myself:

    ./emqtt_bench_sub -c 20000 -i 10 -t bench/%i -q 2
    ./emqtt_bench_pub -c 100 -I 10 -t bench/%i -s 25

Instead of the ```-i 10``` I used ```-i 1``` which allowed to setup the 20000 consumers way faster (I guess 1ms sleep instead of 10ms). No problems occurred during the setup of the consumers. I must mention that I waited until all consumers were started before starting the publishers. After the publishers were publishing I ve seen the expected message rate, however with the given default configuration for ```max_online_messages``` and my pretty old test laptop (with many other programs open) I was constantly dropping some messages (discovered in ```console.log```).

After about 5 minutes I got the mentioned ```{shutdown,tcp_closed}```. Another check in the ```console.log``` file it was clear that VerneMQ has forcefully disconnected those 20k consumers, due to keepalive expired:

    2016-04-24 21:34:26.574 [warning] <0.3967.3>@vmq_mqtt_fsm:connected:389 [<0.3967.3>] stop due to keepalive expired

It actually makes sense that this happened! Although the consumers are subscribed with QoS 2 the messages are delivered as QoS 0. This is because VerneMQ doesn't upgrade the QoS per default (similar to other brokers, not sure im emqttd does though). The ```upgrade_outgoing_qos``` VerneMQ config would allow you to change this behaviour. 
But: it seems that the benchmark subscribers are not sending the required PINGREQs in case of QoS 0.

Regarding my previous comment, I am pretty sure the 'jobs' rate limiter is not an issue here, as the ```emqtt_bench_sub/pub``` is setting up the connections sequentially (at least what I've seen flying over the code). That said, on a decent machine I've seen connection setup rates over 1000 connections per second with VerneMQ, without special tuning of our rate limiter. 

Hope this all makes sense .

Thank you for your suggestion with the dashboard. It hasn't our top priority at the moment but it is on our roadmap.

Cheers,
Andre

**Comment by ioolkos** [comment]
The CONNACK problem is happening for me also. I see TCP-retransmissions of the MQTT Connect Command in Wireshark.

Do you see anything in the VerneMQ console.log?

**Comment by ioolkos** [comment]
The CONNACK problem is isolated to be in LevelDB. A PR with a workaround is pending: https://github.com/erlio/vmq_server/pull/54

**Comment by dcy** [comment]
Thanks!
I will test when the PR merged.

**Comment by liucy2007975** [comment]
@dergraf @ioolkos @dcy  
I did the same test with vernemq and emqttd ,used the vernemq version is 0.12.5p4 .
I test ./emqtt_bench_sub -c 3000 -i 10 -t bench/%i -q 2 for vernemq and emqttd .
after some times , the vernemq console shows message: 
".........@vmq_mqtt_fsm:connected:389 [<0.3967.3>] stop due to keepalive expired " 
and all of the clients stoped by vernemq.

Firstly I thought may be the emqtt_bench_sub would not send keepalive message, Then I test with emqttd, In the console ,I saw , emqtt_bench_sub have send the keepalive message for 5 mins default. 
so the emqtt_bench_sub really have sent the keepalive msg.

as @ioolkos said that " This means your subscriber will disconnect after 5 minutes, if they never issue any MQTT control packets. "  ,Does it  means that the vernemq ignore the keepalive msg as MQTT   control packets .The PINGREQ is also the MQTT   control packets.

@dergraf  said "But: it seems that the benchmark subscribers are not sending the required PINGREQs in case of QoS 0"  from the emqttd console I saw it really sent the PINGREQs . And in vernemq.conf I set upgrade_outgoing_qos=on.  after that the problem always happen.

how to fix that??

**Comment by ioolkos** [comment]
@liucy2007975 

I just did 3 test runs with the exact command you had:  
`./emqtt_bench_sub -c 3000 -i 10 -t bench/%i -q 2`

I observer the traffic in Wireshark. There we see that the connections come in with a keepalive timeout of 300 seconds set.

1) Mosquitto: did not receive a ping req within 300 seconds, and correctly disconnect the connections after 5 min+
2) VerneMQ: did not receive a ping req withing 300 seconds, and correctly disconnect the connections after 5 min+
3) EMQTT did not disconnect the connections, and the emqtt_bench started sending Ping Requests after 600 seconds.

In other words: the bench sends the ping requests too late. Please read what the MQTT spec states re: keep alives, if this still remains unclear.

**Comment by liucy2007975** [comment]
@ioolkos  thank you for your reply!  
Yeah, I see.  and I  just found that the emqttd  detect the clients disconnection time is  (Keepalive * 0.75) .
